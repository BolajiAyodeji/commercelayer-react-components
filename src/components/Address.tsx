import React, {
  FunctionComponent,
  Fragment,
  useContext,
  ReactNode,
  useState,
  useEffect,
} from 'react'
import AddressChildrenContext from '#context/AddressChildrenContext'
import components from '#config/components'
import CustomerContext from '#context/CustomerContext'
import BillingAddressContext from '#context/BillingAddressContext'
import ShippingAddressContext from '#context/ShippingAddressContext'
import { AddressCollection } from '@commercelayer/js-sdk'
import _ from 'lodash'
import AddressContext from '#context/AddressContext'

const propTypes = components.Address.propTypes

type Props = {
  children: ReactNode
  selectedClassName?: string
  onSelect?: () => void
  addresses?: AddressCollection[]
} & JSX.IntrinsicElements['div']

const Address: FunctionComponent<Props> = (props) => {
  const {
    children,
    className,
    selectedClassName,
    onSelect,
    addresses = [],
    ...p
  } = props
  const { addresses: addressesContext } = useContext(CustomerContext)
  const { setBillingAddress, billingCustomerAddressId } = useContext(
    BillingAddressContext
  )
  const { setShippingAddress, shippingCustomerAddressId } = useContext(
    ShippingAddressContext
  )
  const {
    billingAddress,
    shippingAddress,
    shipToDifferentAddress,
  } = useContext(AddressContext)
  const [selected, setSelected] = useState<null | number>(null)
  const items = !_.isEmpty(addresses)
    ? addresses
    : (addressesContext && addressesContext) || []
  useEffect(() => {
    if (items) {
      items.map((address, k) => {
        if (
          billingCustomerAddressId ||
          shippingCustomerAddressId === undefined
        ) {
          const preselected =
            address.customerAddressId === billingCustomerAddressId
          if (preselected && selected === null && _.isEmpty(billingAddress))
            setSelected(k)
          else if (
            (preselected || selected !== null) &&
            !_.isEmpty(billingAddress)
          )
            setSelected(null)
        }
        if (
          (shipToDifferentAddress && shippingCustomerAddressId) ||
          billingCustomerAddressId === undefined
        ) {
          const preselected =
            address.customerAddressId === shippingCustomerAddressId
          if (preselected && selected === null && _.isEmpty(shippingAddress))
            setSelected(k)
          else if (
            (preselected || selected !== null) &&
            !_.isEmpty(shippingAddress)
          )
            setSelected(null)
        }
      })
    }
  }, [billingAddress, shippingAddress, addresses])
  const handleSelect = async (
    k: number,
    addressId: string,
    customerAddressId: string
  ) => {
    if (k !== selected) {
      setSelected(k)
      setBillingAddress &&
        (await setBillingAddress(addressId, { customerAddressId }))
      setShippingAddress &&
        (await setShippingAddress(addressId, { customerAddressId }))
      onSelect && onSelect()
    }
  }
  const components = items.map((address, k) => {
    const addressProps = {
      address,
    }
    const addressSelectedClass =
      selected === k ? `${className} ${selectedClassName}` : className
    const customerAddressId: string = address?.customerAddressId || ''
    return (
      <AddressChildrenContext.Provider key={k} value={addressProps}>
        <div
          className={addressSelectedClass}
          onClick={() => handleSelect(k, address.id, customerAddressId)}
          {...p}
        >
          {children}
        </div>
      </AddressChildrenContext.Provider>
    )
  })
  return <Fragment>{components}</Fragment>
}

Address.propTypes = propTypes

export default Address
