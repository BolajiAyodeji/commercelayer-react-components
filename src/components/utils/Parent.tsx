import PropTypes, { InferProps } from 'prop-types'
import React, { FunctionComponent } from 'react'

const PProps = {
  children: PropTypes.func.isRequired
}

export type ParentProps = InferProps<typeof PProps>

const Parent: FunctionComponent<ParentProps> = props => {
  const Child = props.children
  return <Child {...props} />
}

Parent.propTypes = PProps

export default Parent
