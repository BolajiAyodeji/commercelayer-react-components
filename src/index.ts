export { default as CommerceLayer } from './components/CommerceLayer'
export { default as OrderContainer } from './components/OrderContainer'
export { default as PriceContainer } from './components/PriceContainer'
export { default as Price } from './components/Price'
export { default as VariantContainer } from './components/VariantContainer'
export { default as VariantSelector } from './components/VariantSelector'
export { default as QuantitySelector } from './components/QuantitySelector'
export { default as AddToCart } from './components/AddToCart'
export { default as AvailabilityContainer } from './components/AvailabilityContainer'
export { default as AvailabilityTemplate } from './components/AvailabilityTemplate'
export { default as GiftCardContainer } from './components/GiftCardContainer'
export { default as GiftCard } from './components/GiftCard'
export { default as GiftCardCurrencySelector } from './components/GiftCardCurrencySelector'
export { default as GiftCardInput } from './components/GiftCardInput'
export { default as MetadataInput } from './components/MetadataInput'
export { default as SubmitButton } from './components/SubmitButton'
export { default as ItemContainer } from './components/ItemContainer'
export { default as LineItemsCount } from './components/LineItemsCount'
export { default as LineItemsContainer } from './components/LineItemsContainer'
export { default as LineItem } from './components/LineItem'
export { default as LineItemImage } from './components/LineItemImage'
export { default as LineItemName } from './components/LineItemName'
export { default as LineItemPrice } from './components/LineItemPrice'
export { default as LineItemQuantity } from './components/LineItemQuantity'
export { default as LineItemRemove } from './components/LineItemRemove'
export { default as SubTotal } from './components/SubTotal'
export { default as Discount } from './components/Discount'
export { default as Shipping } from './components/Shipping'
export { default as Taxes } from './components/Taxes'
export { default as GiftCardPrice } from './components/GiftCardPrice'
export { default as Total } from './components/Total'
export { default as Checkout } from './components/Checkout'
export { default as Errors } from './components/Errors'
