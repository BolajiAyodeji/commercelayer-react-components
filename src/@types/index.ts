import { Dispatch } from 'react'
import { BaseError } from '../components/Errors'
import PropTypes, { InferProps, ReactElementLike } from 'prop-types'

export const BC = {
  id: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object
}

export const PTLoader = PropTypes.oneOfType([
  PropTypes.element,
  PropTypes.string
])

export const BaseOrderPricePropsTypes = {
  base: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  children: PropTypes.func,
  format: PropTypes.oneOf(['formatted', 'cents', 'float'])
}

export const BaseOrderComponentPropTypes = {
  children: BaseOrderPricePropsTypes['children'],
  format: BaseOrderPricePropsTypes['format']
}

export type LoaderType = string | ReactElementLike

export const BMObject = PropTypes.objectOf(PropTypes.string)
export type BaseMetadataObject = {
  [key: string]: string | undefined | null
}

export type TimeFormat = 'days' | 'hours'

export type BaseComponent = InferProps<typeof BC>

export interface BaseAction<A = string> {
  type: A
  payload: object
}

export interface BaseState {
  [key: string]: any
  errors?: BaseError[]
}

export type BaseActionType<T = string> = T[]

export interface BFSetStateContainer<T> {
  <P extends T>(param: P): void
}

export interface BaseReducer {
  <S extends BaseState, A extends BaseAction, T extends BaseActionType>(
    state: S,
    action: A,
    type: T
  ): S
}

export interface BaseUnsetState<A> {
  (dispatch: Dispatch<A>): void
}

export interface BaseMetadata {
  [key: string]: string
}

export type InferPropTypes<
  PropTypes,
  DefaultProps = {},
  Props = InferProps<PropTypes>
> = {
  [Key in keyof Props]: Key extends keyof DefaultProps
    ? Props[Key] | DefaultProps[Key]
    : Props[Key]
}
