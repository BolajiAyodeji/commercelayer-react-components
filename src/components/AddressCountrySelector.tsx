import React, { FunctionComponent, useContext, useEffect } from 'react'
import BaseSelect from './utils/BaseSelect'
import countriesOptions from '../config/countries.json'
import components from '../config/components'
import { AddressCountrySelectName, BaseSelectComponentProps } from '../typings'
import BillingAddressContext from '../context/BillingAddressContext'
import ShippingAddressContext from '../context/ShippingAddressContext'

const propTypes = components.AddressCountrySelector.propTypes
const defaultProps = components.AddressCountrySelector.defaultProps
const displayName = components.AddressCountrySelector.displayName

type AddressCountrySelectorProps = Omit<
  BaseSelectComponentProps,
  'options' | 'name'
> & {
  name: AddressCountrySelectName
  required?: boolean
} & Pick<JSX.IntrinsicElements['select'], 'className' | 'id' | 'style'>

const AddressCountrySelector: FunctionComponent<AddressCountrySelectorProps> = (
  props
) => {
  const { required = true, value, name } = props
  const billingAddress = useContext(BillingAddressContext)
  const shippingAddress = useContext(ShippingAddressContext)
  useEffect(() => {
    if (value && billingAddress?.setValue) {
      billingAddress.setValue(name, value)
    }
    if (value && shippingAddress?.setValue) {
      shippingAddress.setValue(name, value)
    }
  }, [value])
  return (
    <BaseSelect
      ref={billingAddress?.validation || shippingAddress?.validation}
      required={required}
      options={countriesOptions}
      {...props}
    />
  )
}

AddressCountrySelector.propTypes = propTypes
AddressCountrySelector.defaultProps = defaultProps
AddressCountrySelector.displayName = displayName

export default AddressCountrySelector
