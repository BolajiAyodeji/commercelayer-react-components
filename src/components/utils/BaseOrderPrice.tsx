import OrderContext from '../../context/OrderContext'
import getAmount from '../../utils/getAmount'
import Parent from './Parent'
import React, {
  FunctionComponent,
  useState,
  useEffect,
  useContext
} from 'react'
import PropTypes, { InferProps } from 'prop-types'

export const BOPProps = {
  base: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  children: PropTypes.func,
  format: PropTypes.oneOf(['formatted', 'cents', 'float'])
}

export const BOCProps = {
  children: BOPProps['children'],
  format: BOPProps['format']
}

export type BaseOrderComponentProps = InferProps<typeof BOCProps>

export type BaseOrderPriceProps = InferProps<typeof BOPProps> &
  JSX.IntrinsicElements['span']

const BaseOrderPrice: FunctionComponent<BaseOrderPriceProps> = props => {
  const { format, base, type, children, ...p } = props
  const { order } = useContext(OrderContext)
  const [price, setPrice] = useState('')
  useEffect(() => {
    const p = getAmount(base, type, format || 'formatted', order)
    setPrice(p)
    return (): void => {
      setPrice('')
    }
  }, [order])
  const parentProps = {
    price,
    ...p
  }
  return props.children ? (
    <Parent {...parentProps}>{children}</Parent>
  ) : (
    <span {...p}>{price}</span>
  )
}

BaseOrderPrice.propTypes = BOPProps

BaseOrderPrice.defaultProps = {
  format: 'formatted'
}

export default BaseOrderPrice
