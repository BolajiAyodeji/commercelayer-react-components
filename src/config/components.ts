import PropTypes from 'prop-types'
import childrenTypes from '../utils/childrenTypes'
import {
  TimeFormat,
  BaseInputType,
  GiftCardInputName,
  baseOrderComponentPricePropTypes,
  LineItemType,
  BasePriceType,
  BaseFormatPrice,
  PTLoader,
  BMObject,
  BaseSelectorType,
} from '../@types'
import { ErrorPropTypes } from '../@types/errors'
import { BaseInputComponentPropTypes } from '../@types/index'

const components = {
  AddToCart: {
    displayName: 'AddToCart',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
      skuCode: PropTypes.string,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      label: 'Add to cart',
    },
  },
  AvailabilityContainer: {
    displayName: 'AvailabilityContainer',
    description: '',
    permittedChildren: ['AvailabilityTemplate', 'ReactNode'],
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  AvailabilityTemplate: {
    displayName: 'AvailabilityTemplate',
    description: '',
    propTypes: {
      timeFormat: PropTypes.oneOf<TimeFormat>(['days', 'hours']),
      showShippingMethodName: PropTypes.bool,
      children: PropTypes.func,
    },
    defaultProps: {
      timeFormat: 'days',
      showShippingMethodName: false,
    },
  },
  Checkout: {
    displayName: 'Checkout',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'Checkout',
    },
  },
  CommerceLayer: {
    permittedChildren: [
      'OrderContainer',
      'PriceContainer',
      'GiftCardContainer',
      'ReactNode',
    ],
    propTypes: {
      children: childrenTypes.isRequired,
      accessToken: PropTypes.string.isRequired,
      endpoint: PropTypes.string.isRequired,
      cache: PropTypes.bool,
    },
    defaultProps: {
      cache: false,
    },
  },
  Discount: {
    displayName: 'Discount',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Errors: {
    displayName: 'Errors',
    description: '',
    propTypes: ErrorPropTypes,
    defaultProps: {
      messages: [],
      field: 'base',
    },
  },
  GiftCard: {
    permittedChildren: [
      'GiftCardCurrencySelector',
      'GiftCardInput',
      'Errors',
      'MetadataInput',
      'SubmitButton',
      'ReactNode',
    ],
    displayName: 'GiftCard',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      metadata: PropTypes.objectOf(PropTypes.string),
      onSubmit: PropTypes.func,
    },
    defaultProps: {
      onSubmit: undefined,
      metadata: {},
    },
  },
  GiftCardContainer: {
    permittedChildren: ['GiftCard', 'ReactNode'],
    displayName: 'GiftCardContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
    },
  },
  GiftCardCurrencySelector: {
    displayName: 'GiftCardCurrencySelector',
    description: '',
    propTypes: {
      children: PropTypes.func,
      placeholder: PropTypes.exact({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        label: PropTypes.string,
      }),
      value: PropTypes.string,
      required: PropTypes.bool,
    },
    defaultProps: {
      required: true,
    },
  },
  GiftCardInput: {
    displayName: 'GiftCardInput',
    description: '',
    propTypes: {
      type: PropTypes.oneOf<BaseInputType>([
        'text',
        'email',
        'number',
        'date',
        'checkbox',
      ]).isRequired,
      name: PropTypes.oneOf<GiftCardInputName>([
        'balanceCents',
        'balanceMaxCents',
        'singleUse',
        'rechargeable',
        'imageUrl',
        'expiresAt',
        'referenceOrigin',
        'email',
        'firstName',
        'lastName',
        'reference',
      ]).isRequired,
      children: PropTypes.func,
      placeholder: PropTypes.string,
    },
  },
  GiftCardPrice: {
    displayName: 'GiftCardPrice',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
  },
  ItemContainer: {
    permittedChildren: [
      'PriceContainer',
      'VariantContainer',
      'SkuOptionContainer',
      'QuantitySelector',
      'AddToCart',
      'AvailabilityContainer',
      'ReactNode',
    ],
    displayName: 'ItemContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  LineItem: {
    permittedChildren: [
      'LineItemImage',
      'LineItemName',
      'LineItemOptions',
      'LineItemQuantity',
      'LineItemPrice',
      'LineItemRemove',
      'Errors',
      'ReactNode',
    ],
    displayName: 'LineItem',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      type: PropTypes.oneOf<LineItemType>([
        'skus',
        'gift_cards',
        'shipments',
        'paymentMethods',
        'promotions',
      ]),
    },
    defaultProps: {
      type: 'skus',
    },
  },
  LineItemImage: {
    displayName: 'LineItemImage',
    description: '',
    propTypes: {
      width: PropTypes.number,
      src: PropTypes.string,
      children: PropTypes.func,
    },
  },
  LineItemName: {
    displayName: 'LineItemName',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
  },
  LineItemOption: {
    displayName: 'LineItemOption',
    description: '',
    propTypes: {
      name: PropTypes.string.isRequired,
      children: PropTypes.func,
      valueClassName: PropTypes.string,
      keyClassName: PropTypes.string,
      keyId: PropTypes.string,
      keyStyle: PropTypes.object,
    },
  },
  LineItemOptions: {
    permittedChildren: ['LineItemOption', 'ReactNode'],
    displayName: 'LineItemOptions',
    description: '',
    propTypes: {
      name: PropTypes.string.isRequired,
      children: childrenTypes.isRequired,
      showName: PropTypes.bool,
    },
    defaultProps: {
      showName: true,
    },
  },
  LineItemPrice: {
    displayName: 'LineItemPrice',
    description: '',
    propTypes: {
      ...baseOrderComponentPricePropTypes,
      type: PropTypes.oneOf<BasePriceType>(['total', 'unit', 'option']),
    },
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
      type: 'total' as BasePriceType,
    },
  },
  LineItemQuantity: {
    displayName: 'LineItemQuantity',
    description: '',
    propTypes: {
      children: PropTypes.func,
      max: PropTypes.number,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      max: 50,
    },
  },
  LineItemRemove: {
    displayName: 'LineItemRemove',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'remove',
    },
  },
  LineItemsContainer: {
    permittedChildren: ['LineItemsCount', 'LineItem', 'ReactNode'],
    displayName: 'LineItemsContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      filters: PropTypes.object,
      loader: PTLoader,
    },
    defaultProps: {
      filters: {},
      loader: 'Loading...',
    },
  },
  LineItemsCount: {
    displayName: 'LineItemsCount',
    description: '',
    propTypes: {
      children: PropTypes.func,
      id: PropTypes.string,
      className: PropTypes.string,
      name: PropTypes.string,
      style: PropTypes.object,
    },
  },
  MetadataInput: {
    displayName: 'MetadataInput',
    description: '',
    propTypes: BaseInputComponentPropTypes,
  },
  OrderContainer: {
    permittedChildren: [
      'ItemContainer',
      'LineItemsContainer',
      'SubTotal',
      'Discount',
      'Shipping',
      'Taxes',
      'GiftCardPrice',
      'Total',
      'Checkout',
      'GiftCardContainer',
      'ReactNode',
    ],
    displayName: 'OrderContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      persistKey: PropTypes.string.isRequired,
      metadata: BMObject,
    },
    defaultProps: {
      metadata: {},
    },
  },
  Price: {
    displayName: 'Price',
    description: '',
    propTypes: {
      children: PropTypes.func,
      compareClassName: PropTypes.string,
      skuCode: PropTypes.string,
      showCompare: PropTypes.bool,
    },
    defaultProps: {
      skuCode: '',
    },
  },
  PriceContainer: {
    permittedChildren: ['Price', 'ReactNode'],
    displayName: 'PriceContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
      loader: PTLoader,
      perPage: PropTypes.number,
      filters: PropTypes.object,
    },
    defaultProps: {
      perPage: 10,
      filters: {},
      loader: 'Loading...',
      skuCode: '',
    },
  },
  PriceTemplate: {
    displayName: 'PriceTemplate',
    description: '',
    propTypes: {
      formattedAmount: PropTypes.string,
      formattedCompare: PropTypes.string,
    },
    defaultProps: {
      formattedAmount: '',
      formattedCompare: '',
    },
  },
  QuantitySelector: {
    displayName: 'QuantitySelector',
    description: '',
    propTypes: {
      children: PropTypes.func,
      min: PropTypes.string,
      max: PropTypes.string,
      value: PropTypes.string,
      skuCode: PropTypes.string,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      min: '1',
    },
  },
  Shipping: {
    displayName: 'Shipping',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
  },
  SkuOption: {
    permittedChildren: ['SkuOptionInput', 'ReactNode'],
    displayName: 'SkuOption',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      name: PropTypes.string.isRequired,
    },
  },
  SkuOptionInput: {
    displayName: 'SkuOptionInput',
    description: '',
    propTypes: BaseInputComponentPropTypes,
  },
  SkuOptionsContainer: {
    permittedChildren: ['SkuOption', 'ReactNode'],
    displayName: 'SkuOptionsContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  SubmitButton: {
    displayName: 'SubmitButton',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'Submit',
    },
  },
  SubTotal: {
    displayName: 'SubTotal',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Taxes: {
    displayName: 'Taxes',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Total: {
    displayName: 'Total',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  VariantContainer: {
    permittedChildren: ['VariantSelector', 'ReactNode'],
    displayName: 'VariantContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
      filters: PropTypes.object,
    },
    defaultProps: {
      skuCode: '',
      filters: {},
    },
  },
  VariantSelector: {
    displayName: 'VariantSelector',
    description: '',
    propTypes: {
      skuCodes: PropTypes.arrayOf(
        PropTypes.exact({
          label: PropTypes.string.isRequired,
          code: PropTypes.string.isRequired,
        }).isRequired
      ).isRequired,
      name: PropTypes.string,
      children: PropTypes.func,
      type: PropTypes.oneOf<BaseSelectorType>(['select', 'radio']),
      loader: PropTypes.element,
      placeholder: PropTypes.string,
      skuCode: PropTypes.string,
    },
    defaultProps: {
      placeholder: 'select an variant',
      type: 'select' as BaseSelectorType,
    },
  },
  VariantTemplate: {
    displayName: 'VariantTemplate',
    description: '',
    propTypes: {
      variants: PropTypes.object.isRequired,
      onChange: PropTypes.func,
      skuCodes: PropTypes.arrayOf(
        PropTypes.exact({
          label: PropTypes.string.isRequired,
          code: PropTypes.string.isRequired,
        }).isRequired
      ).isRequired,
      name: PropTypes.string,
      children: PropTypes.func,
      type: PropTypes.oneOf<BaseSelectorType>(['select', 'radio']),
      loader: PropTypes.element,
      placeholder: PropTypes.string,
      skuCode: PropTypes.string,
    },
  },
}

export default components
