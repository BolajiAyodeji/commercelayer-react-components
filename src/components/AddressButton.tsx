import React, { FunctionComponent, useContext } from 'react'
import Parent from './utils/Parent'
import components from '../config/components'
import { FunctionChildren } from '../typings/index'
import AddressContext from '../context/AddressContext'
import _ from 'lodash'
import { fieldsExist } from '../utils/validateFormFields'
import { addressFields } from '../reducers/AddressReducer'

const propTypes = components.AddressButton.propTypes
const defaultProps = components.AddressButton.defaultProps
const displayName = components.AddressButton.displayName

type AddressButtonChildrenProps = FunctionChildren<
  Omit<AddressButtonProps, 'children'>
>

type AddressButtonProps = {
  children?: AddressButtonChildrenProps
  label?: string
  onClick?: () => void
} & JSX.IntrinsicElements['button']

const AddressButton: FunctionComponent<AddressButtonProps> = (props) => {
  const {
    children,
    label = 'Continue to delivery',
    resource,
    disabled,
    onClick,
    ...p
  } = props
  const {
    errors,
    billingAddress,
    shipToDifferentAddress,
    shippingAddress,
    saveAddresses,
  } = useContext(AddressContext)
  let disable =
    disabled ||
    !_.isEmpty(errors) ||
    _.isEmpty(billingAddress) ||
    (shipToDifferentAddress && _.isEmpty(shippingAddress))
  if (_.isEmpty(errors) && !_.isEmpty(billingAddress)) {
    disable = billingAddress && fieldsExist(billingAddress)
  }
  if (
    _.isEmpty(errors) &&
    shipToDifferentAddress
    // !_.isEmpty(shippingAddress)
  ) {
    disable =
      shippingAddress &&
      fieldsExist(shippingAddress, _.without(addressFields, 'customer_email'))
  }
  const handleClick = () => {
    if (_.isEmpty(errors) && !disable) {
      saveAddresses()
      onClick && onClick()
    }
  }
  const parentProps = {
    ...p,
    label,
    resource,
    handleClick,
    disabled: disable,
  }
  return children ? (
    <Parent {...parentProps}>{children}</Parent>
  ) : (
    <button type="button" disabled={disable} onClick={handleClick} {...p}>
      {label}
    </button>
  )
}

AddressButton.propTypes = propTypes
AddressButton.defaultProps = defaultProps
AddressButton.displayName = displayName

export default AddressButton
