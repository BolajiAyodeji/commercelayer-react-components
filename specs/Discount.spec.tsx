import React from 'react'
import { Discount } from '../src'
import renderer from 'react-test-renderer'
import components from '../src/config/components'
import BaseOrderPrice from '../src/components/utils/BaseOrderPrice'
import Parent from '../src/components/utils/Parent'

const propTypes = components.Discount.propTypes

test('<Discount/>', () => {
  expect.assertions(3)
  const component = renderer.create(<Discount />)
  const tree = component.toJSON()
  const root = component.toTree()
  const proptypes = root.type['propTypes']
  expect(tree).toMatchSnapshot()
  expect(proptypes.children).toBe(propTypes.children)
  expect(proptypes.format).toBe(propTypes.format)
})

test('<Discount children rendered />', () => {
  expect.assertions(3)
  const component = renderer.create(<Discount />)
  const tree = component.toJSON()
  const root = component.toTree()
  const rendered = root.rendered
  expect(tree).toMatchSnapshot()
  expect(rendered.nodeType).toBe('component')
  expect(rendered.type).toBe(BaseOrderPrice)
})

test('<Discount with custom children />', () => {
  expect.assertions(5)
  const CustomComponent = props => <span>{props.label}</span>
  const component = renderer.create(
    <Discount format="cents">{CustomComponent}</Discount>
  )
  const tree = component.toJSON()
  const root = component.toTree()
  const rendered = root.rendered
  const parentRendered = root.rendered.rendered
  const childRendered = parentRendered.rendered
  expect(tree).toMatchSnapshot()

  expect(rendered.props.children).toBe(CustomComponent)

  expect(parentRendered.nodeType).toBe('component')
  expect(parentRendered.type).toBe(Parent)

  expect(childRendered.props.price).not.toBeUndefined()
})
