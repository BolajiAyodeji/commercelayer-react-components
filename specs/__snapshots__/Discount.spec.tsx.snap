// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Discount children rendered /> 1`] = `
<span>
  
</span>
`;

exports[`<Discount with custom children /> 1`] = `<span />`;

exports[`<Discount/> 1`] = `
<span>
  
</span>
`;
